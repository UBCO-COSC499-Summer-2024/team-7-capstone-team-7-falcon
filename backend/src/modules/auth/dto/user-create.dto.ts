import {
  IsEmail,
  IsInt,
  IsNotEmpty,
  IsOptional,
  IsString,
  IsStrongPassword,
  Length,
  Min,
} from 'class-validator';
import { ERROR_MESSAGES } from '../../../common';
import { MatchPasswords } from '../../../decorators/match-passwords.decorator';

export class UserCreateDto {
  @IsString({ message: ERROR_MESSAGES.userEditDto.firstNameString })
  @IsNotEmpty({ message: ERROR_MESSAGES.userEditDto.firstNameRequired })
  @Length(2, 15, { message: ERROR_MESSAGES.userEditDto.firstNameLength })
  readonly first_name!: string;

  @IsOptional()
  readonly last_name?: string;

  @IsEmail()
  @IsNotEmpty({ message: ERROR_MESSAGES.common.emailRequired })
  readonly email!: string;

  @IsString({ message: ERROR_MESSAGES.common.passwordString })
  @IsNotEmpty({ message: ERROR_MESSAGES.common.passwordRequired })
  @IsStrongPassword(
    {
      minLength: 8,
      minLowercase: 1,
      minUppercase: 1,
      minNumbers: 1,
      minSymbols: 1,
    },
    { message: ERROR_MESSAGES.common.passwordStrong },
  )
  readonly password!: string;

  @IsString({ message: ERROR_MESSAGES.common.confirmPasswordString })
  @IsNotEmpty({ message: ERROR_MESSAGES.common.confirmPasswordRequired })
  @MatchPasswords('password', {
    message: ERROR_MESSAGES.common.passwordMustMatch,
  })
  readonly confirm_password!: string;

  @IsInt({ message: ERROR_MESSAGES.userEditDto.studentNumberMustBeNumber })
  @Min(1, { message: ERROR_MESSAGES.userEditDto.studentNumberMustBeGreater })
  @IsOptional()
  readonly student_id?: number;

  @IsInt({ message: ERROR_MESSAGES.userEditDto.employeeNumberMustBeNumber })
  @Min(1, { message: ERROR_MESSAGES.userEditDto.employeeNumberMustBeGreater })
  @IsOptional()
  readonly employee_id?: number;
}
