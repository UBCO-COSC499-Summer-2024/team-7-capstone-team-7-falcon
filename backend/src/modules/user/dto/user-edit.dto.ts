import { IsInt, IsOptional, IsString, Length, Min } from 'class-validator';
import { ERROR_MESSAGES } from '../../../common';

export class UserEditDto {
  @IsString({ message: ERROR_MESSAGES.userEditDto.firstNameString })
  @Length(2, 15, { message: ERROR_MESSAGES.userEditDto.firstNameLength })
  @IsOptional()
  readonly first_name?: string;

  @IsOptional()
  readonly last_name?: string;

  @IsInt({ message: ERROR_MESSAGES.userEditDto.studentNumberMustBeNumber })
  @Min(1, { message: ERROR_MESSAGES.userEditDto.studentNumberMustBeGreater })
  @IsOptional()
  readonly student_id?: number;

  @IsInt({ message: ERROR_MESSAGES.userEditDto.employeeNumberMustBeNumber })
  @Min(1, { message: ERROR_MESSAGES.userEditDto.employeeNumberMustBeGreater })
  @IsOptional()
  readonly employee_id?: number;
}
