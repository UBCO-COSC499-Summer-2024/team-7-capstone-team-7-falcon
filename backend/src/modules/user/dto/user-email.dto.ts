import { IsEmail, IsNotEmpty } from 'class-validator';
import { ERROR_MESSAGES } from '../../../common';

export class UserEmailDto {
  @IsEmail({}, { message: ERROR_MESSAGES.common.emailInvalid })
  @IsNotEmpty({ message: ERROR_MESSAGES.common.emailRequired })
  readonly email!: string;
}
