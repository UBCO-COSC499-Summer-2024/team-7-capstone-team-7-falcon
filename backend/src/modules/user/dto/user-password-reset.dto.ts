import { IsNotEmpty, IsString, IsStrongPassword } from 'class-validator';
import { ERROR_MESSAGES } from '../../../common';
import { MatchPasswords } from '../../../decorators/match-passwords.decorator';

export class UserPasswordResetDto {
  @IsString({ message: ERROR_MESSAGES.tokenController.tokenString })
  @IsNotEmpty({ message: ERROR_MESSAGES.tokenController.tokenRequired })
  readonly token!: string;

  @IsString({ message: ERROR_MESSAGES.common.passwordString })
  @IsNotEmpty({ message: ERROR_MESSAGES.common.passwordRequired })
  @IsStrongPassword(
    {
      minLength: 8,
      minLowercase: 1,
      minUppercase: 1,
      minNumbers: 1,
      minSymbols: 1,
    },
    { message: ERROR_MESSAGES.common.passwordStrong },
  )
  readonly password!: string;

  @IsString({ message: ERROR_MESSAGES.common.confirmPasswordString })
  @IsNotEmpty({ message: ERROR_MESSAGES.common.confirmPasswordRequired })
  @MatchPasswords('password', {
    message: ERROR_MESSAGES.common.passwordMustMatch,
  })
  readonly confirm_password!: string;
}
