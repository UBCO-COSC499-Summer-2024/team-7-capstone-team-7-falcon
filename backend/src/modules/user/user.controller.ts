import { Controller, HttpStatus, ParseIntPipe } from '@nestjs/common';
import { Get, Param, Res } from '@nestjs/common/decorators';
import { Response } from 'express';
import { UserService } from './user.service';
import { ERROR_MESSAGES } from '../../common';

@Controller('user')
export class UserController {
  constructor(private readonly userService: UserService) {}

  @Get('/')
  get(@Res() res: Response) {
    return res.status(HttpStatus.OK).send({
      message: 'ok',
    });
  }

  /**
   * Get user by id
   * @param res {Response} - Response object
   * @param uid {number} - User id
   * @returns {Promise<Response>} - User object
   */
  @Get('/:uid')
  async getById(
    @Res() res: Response,
    @Param('uid', ParseIntPipe) uid: number,
  ): Promise<Response> {
    const user = await this.userService.getUserById(uid);

    if (!user) {
      return res.status(HttpStatus.NOT_FOUND).send({
        message: ERROR_MESSAGES.userController.userNotFound,
      });
    } else {
      return res.status(HttpStatus.OK).send(user);
    }
  }
}
