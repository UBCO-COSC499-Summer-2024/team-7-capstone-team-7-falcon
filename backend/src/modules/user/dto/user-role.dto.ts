import { IsEnum, IsNotEmpty } from 'class-validator';
import { ERROR_MESSAGES } from '../../../common';
import { UserRoleEnum } from '../../../enums/user.enum';

export class UserRoleDto {
  @IsEnum(UserRoleEnum, { message: ERROR_MESSAGES.userRoleDto.userRoleInvalid })
  @IsNotEmpty({ message: ERROR_MESSAGES.userRoleDto.userRoleRequired })
  userRole!: UserRoleEnum;
}
