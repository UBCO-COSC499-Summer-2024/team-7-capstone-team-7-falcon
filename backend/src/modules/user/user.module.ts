import { Module, forwardRef } from '@nestjs/common';
import { UserController } from './user.controller';
import { UserService } from './user.service';
import { TokenModule } from '../token/token.module';

@Module({
  controllers: [UserController],
  providers: [UserService],
  exports: [UserService],
  imports: [forwardRef(() => TokenModule)],
})
export class UserModule {}
