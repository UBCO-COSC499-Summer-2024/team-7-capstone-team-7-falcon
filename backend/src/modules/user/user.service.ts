import { Injectable } from '@nestjs/common';
import { UserModel } from './entities/user.entity';
import { OAuthGoogleUserPayload } from '../../common/interfaces';
import {
  EmployeeIdAlreadyExistsException,
  StudentIdAlreadyExistsException,
  UserAlreadyExistsException,
  UserNotFoundException,
} from '../../common/errors';
import { AuthTypeEnum } from '../../enums/user.enum';
import { EmployeeUserModel } from './entities/employee-user.entity';
import { StudentUserModel } from './entities/student-user.entity';
import { UserEditDto } from './dto/user-edit.dto';
import { CourseUserModel } from '../course/entities/course-user.entity';

@Injectable()
export class UserService {
  /**
   * Search for courses based on user id
   * @param user_id {number} - User id
   * @returns {Promise<CourseUserModel[]>} - CourseUserModel[] promise
   */
  public async findUserCoursesById(
    user_id: number,
  ): Promise<CourseUserModel[]> {
    const courses: CourseUserModel[] = await CourseUserModel.find({
      where: { user: { id: user_id }, course: { is_archived: false } },
      relations: ['user', 'course'],
    });

    if (!courses || courses.length === 0) {
      return null;
    }

    return courses;
  }

  /**
   * Returns a user by id
   * @param id {number} - User id
   * @returns {Promise<UserModel>} - User object
   */
  public async getUserById(id: number): Promise<UserModel> {
    const user: UserModel = await UserModel.findOne({
      where: { id },
      relations: ['employee_user', 'student_user'],
    });

    if (!user) {
      return null;
    }

    // Remove password from user object as it is sensitive information
    delete user.password;

    return user;
  }

  /**
   * Returns a user by email
   * @param email {string} - User email
   * @returns {Promise<UserModel>} - User object
   */
  public async findOrCreateUser(
    userPayload: OAuthGoogleUserPayload,
    method: string,
  ): Promise<UserModel> {
    if (method === AuthTypeEnum.GOOGLE_OAUTH) {
      let user: UserModel = await UserModel.findOne({
        where: { email: userPayload.email },
      });

      if (!user) {
        const currentTime = parseInt(new Date().getTime().toString());

        user = new UserModel();
        user.email = userPayload.email;
        user.first_name = userPayload.given_name;
        user.last_name = userPayload.family_name;
        user.avatar_url = userPayload.picture;
        user.auth_type = AuthTypeEnum.GOOGLE_OAUTH;
        user.created_at = currentTime;
        user.updated_at = currentTime;

        await user.save();
      }

      if (user && user.auth_type !== AuthTypeEnum.GOOGLE_OAUTH) {
        throw new UserAlreadyExistsException();
      }

      return user;
    }

    throw new Error('Invalid auth method');
  }

  /**
   * Edit user details
   * @param uid {number} - User id
   * @param userEditBody {UserEditDto} - User edit body
   * @returns {Promise<UserModel>} - User object
   */
  public async editUser(
    uid: number,
    userEditBody: UserEditDto,
  ): Promise<UserModel> {
    const user: UserModel = await UserModel.findOne({
      where: { id: uid },
      relations: ['employee_user', 'student_user'],
    });

    if (!user) {
      throw new UserNotFoundException();
    }

    if (userEditBody.first_name) {
      user.first_name = userEditBody.first_name;
    }

    if (userEditBody.last_name) {
      user.last_name = userEditBody.last_name;
    }

    // Create or update employee number
    if (userEditBody.employee_id) {
      const employeeUser: EmployeeUserModel = await this.findEmployeeNumber(
        userEditBody.employee_id,
      );

      if (employeeUser && employeeUser.user.id !== uid) {
        throw new EmployeeIdAlreadyExistsException();
      }

      let employeeUserRecord: EmployeeUserModel;

      if (!user.employee_user) {
        employeeUserRecord = await EmployeeUserModel.create({
          employee_id: userEditBody.employee_id,
          user,
        }).save();
        user.employee_user = employeeUserRecord;
      } else {
        user.employee_user.employee_id = userEditBody.employee_id;
        await user.employee_user.save();
      }
    }

    // Create or update student number
    if (userEditBody.student_id) {
      const studentUser: StudentUserModel = await this.findStudentNumber(
        userEditBody.student_id,
      );

      if (studentUser && studentUser.user.id !== uid) {
        throw new StudentIdAlreadyExistsException();
      }

      let studentUserRecord: StudentUserModel;

      if (!user.student_user) {
        studentUserRecord = await StudentUserModel.create({
          student_id: userEditBody.student_id,
          user,
        }).save();
        user.student_user = studentUserRecord;
        user.student_user.save();
      } else {
        user.student_user.student_id = userEditBody.student_id;
        user.student_user.save();
      }
    }
    await user.save();

    return user;
  }

  /**
   * Checks if an employee number exists in the database
   * @param employeeId {number} - Employee number
   * @param userId {number} - User id
   * @returns {Promise<EmployeeUserModel>} - Employee user object
   */
  private async findEmployeeNumber(
    employeeId: number,
  ): Promise<EmployeeUserModel> {
    return await EmployeeUserModel.findOne({
      where: { employee_id: employeeId },
      relations: ['user'],
    });
  }

  /**
   * Checks if a student number exists in the database
   * @param studentId {number} - Student number
   * @returns {Promise<StudentUserModel>} - Student user object
   */
  private async findStudentNumber(
    studentId: number,
  ): Promise<StudentUserModel> {
    return await StudentUserModel.findOne({
      where: { student_id: studentId },
      relations: ['user'],
    });
  }
}
