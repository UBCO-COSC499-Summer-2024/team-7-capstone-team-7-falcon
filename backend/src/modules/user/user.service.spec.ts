import { Test, TestingModule } from '@nestjs/testing';
import { UserService } from './user.service';
import {
  TestConfigModule,
  TestTypeOrmModule,
} from '../../../test/utils/testUtils';
import { UserModel } from './entities/user.entity';
import { EmployeeUserModel } from './entities/employee-user.entity';
import { StudentUserModel } from './entities/student-user.entity';
import { AuthTypeEnum, CourseRoleEnum } from '../../enums/user.enum';
import { OAuthGoogleUserPayload } from '../../common/interfaces';
import { faker } from '@faker-js/faker';
import { CourseModel } from '../course/entities/course.entity';
import { CourseUserModel } from '../course/entities/course-user.entity';
import { TokenModule } from '../token/token.module';
import { UserCreateDto } from '../auth/dto/user-create.dto';

describe('UserService', () => {
  let userService: UserService;
  let moduleRef: TestingModule;

  beforeEach(async () => {
    moduleRef = await Test.createTestingModule({
      providers: [UserService],
      imports: [TestTypeOrmModule, TestConfigModule, TokenModule],
    }).compile();

    userService = moduleRef.get<UserService>(UserService);
  });

  afterEach(async () => {
    await moduleRef.close();
  });

  describe('getUserById', () => {
    it('should return user', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await EmployeeUserModel.create({
        employee_id: 1,
        user: user,
      }).save();

      await StudentUserModel.create({
        student_id: 1,
        user: user,
      }).save();

      const result = await userService.getUserById(user.id);
      expect(result).toMatchSnapshot();
    });

    it("should return null when user id doesn't exist", async () => {
      const user = await userService.getUserById(100);
      expect(user).toBeNull();
    });
  });

  describe('findOrCreateUser', () => {
    it('should create user', async () => {
      const userPayload: OAuthGoogleUserPayload = {
        email: 'john.doe@test.com',
        given_name: 'John',
        family_name: 'Doe',
        picture: 'picture',
      };

      const user = await userService.findOrCreateUser(
        userPayload,
        AuthTypeEnum.GOOGLE_OAUTH,
      );

      delete user.created_at;
      delete user.updated_at;

      expect(user).toMatchSnapshot();
    });

    it('should throw an error when user already exists with different auth type', async () => {
      const userPayload: OAuthGoogleUserPayload = {
        email: 'john.doe@test.com',
        given_name: 'John',
        family_name: 'Doe',
        picture: 'picture',
      };

      await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await expect(
        userService.findOrCreateUser(userPayload, AuthTypeEnum.GOOGLE_OAUTH),
      ).rejects.toThrow('User already exists');
    });

    it('should throw an error when user already exists with the same auth type for EMAIL AccountType', async () => {
      const password = faker.internet.password();
      const email = faker.internet.email();

      await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: email,
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const userPayload: UserCreateDto = {
        first_name: 'John',
        last_name: 'Doe',
        email: email,
        password: password,
        confirm_password: password,
      };

      await expect(
        userService.findOrCreateUser(userPayload, AuthTypeEnum.EMAIL),
      ).rejects.toThrow('User already exists');
    });

    it('should throw an error when user does not have student or employee field for EMAIL AccountType', async () => {
      const password = faker.internet.password();
      const email = faker.internet.email();

      const userPayload: UserCreateDto = {
        first_name: 'John',
        last_name: 'Doe',
        email: email,
        password: password,
        confirm_password: password,
      };

      await expect(
        userService.findOrCreateUser(userPayload, AuthTypeEnum.EMAIL),
      ).rejects.toThrow('Student or employee ID fields are missing');
    });

    it('should throw an error when employee id already exists for EMAIL AccountType', async () => {
      const password = faker.internet.password();
      const email = faker.internet.email();

      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@email.com',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await EmployeeUserModel.create({
        employee_id: 1,
        user: user,
      }).save();

      const userPayload: UserCreateDto = {
        first_name: 'John',
        last_name: 'Doe',
        email: email,
        password: password,
        confirm_password: password,
        employee_id: 1,
      };

      await expect(
        userService.findOrCreateUser(userPayload, AuthTypeEnum.EMAIL),
      ).rejects.toThrow('Employee ID already exists');
    });

    it('should throw an error when student id already exists for EMAIL AccountType', async () => {
      const password = faker.internet.password();
      const email = faker.internet.email();

      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@email.com',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await StudentUserModel.create({
        student_id: 1,
        user: user,
      }).save();

      const userPayload: UserCreateDto = {
        first_name: 'John',
        last_name: 'Doe',
        email: email,
        password: password,
        confirm_password: password,
        student_id: 1,
      };

      await expect(
        userService.findOrCreateUser(userPayload, AuthTypeEnum.EMAIL),
      ).rejects.toThrow('Student ID already exists');
    });

    it('should create user with email auth type and student id', async () => {
      const password = 'password';
      const email = 'email@email.com';

      const userPayload: UserCreateDto = {
        first_name: 'John',
        last_name: 'Doe',
        email: email,
        password: password,
        confirm_password: password,
        student_id: 1,
      };

      const user = await userService.findOrCreateUser(
        userPayload,
        AuthTypeEnum.EMAIL,
      );

      delete user.created_at;
      delete user.updated_at;
      delete user.password;
      delete user.tokens[0].created_at;
      delete user.tokens[0].expires_at;
      delete user.tokens[0].token;
      delete user.student_user.user;

      expect(user).toMatchSnapshot();
    });

    it('should create user with email auth type and employee id', async () => {
      const password = 'password';
      const email = 'email@email.com';

      const userPayload: UserCreateDto = {
        first_name: 'John',
        last_name: 'Doe',
        email: email,
        password: password,
        confirm_password: password,
        employee_id: 1,
      };

      const user = await userService.findOrCreateUser(
        userPayload,
        AuthTypeEnum.EMAIL,
      );

      delete user.created_at;
      delete user.updated_at;
      delete user.password;
      delete user.tokens[0].created_at;
      delete user.tokens[0].expires_at;
      delete user.tokens[0].token;
      delete user.employee_user.user;

      expect(user).toMatchSnapshot();
    });

    it('should create user with email auth type and student id and employee id', async () => {
      const password = 'password';
      const email = 'email@email.com';

      const userPayload: UserCreateDto = {
        first_name: 'John',
        last_name: 'Doe',
        email: email,
        password: password,
        confirm_password: password,
        employee_id: 1,
        student_id: 1,
      };

      const user = await userService.findOrCreateUser(
        userPayload,
        AuthTypeEnum.EMAIL,
      );

      delete user.created_at;
      delete user.updated_at;
      delete user.password;
      delete user.tokens[0].created_at;
      delete user.tokens[0].expires_at;
      delete user.tokens[0].token;
      delete user.employee_user.user;
      delete user.student_user.user;

      expect(user).toMatchSnapshot();
    });

    it('should throw an error when auth type is not supported', async () => {
      const userPayload: OAuthGoogleUserPayload = {
        email: 'john.doe@test.com',
        given_name: 'John',
        family_name: 'Doe',
        picture: 'picture',
      };

      await expect(
        userService.findOrCreateUser(userPayload, 'invalid_auth_type'),
      ).rejects.toThrow('Invalid auth method');
    });
  });

  describe('editUser', () => {
    it('should throw an error when user is not found', async () => {
      await expect(
        userService.editUser(100, {
          first_name: faker.person.firstName(),
        }),
      ).rejects.toThrow('User not found');
    });

    it('should edit user first name', async () => {
      const oldFirstName = faker.person.firstName();

      await UserModel.create({
        first_name: oldFirstName,
        last_name: faker.person.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const newFirstName = faker.person.firstName();

      const user = await userService.editUser(1, {
        first_name: newFirstName,
      });

      expect(user.first_name).toBe(newFirstName);
    });

    it('should edit user last name', async () => {
      const oldLastName = faker.person.lastName();
      const firstName = faker.person.firstName();

      await UserModel.create({
        first_name: firstName,
        last_name: oldLastName,
        email: faker.internet.email(),
        password: faker.internet.password(),
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const newLastName = faker.person.lastName();

      const user = await userService.editUser(1, {
        first_name: firstName,
        last_name: newLastName,
      });

      expect(user.last_name).toBe(newLastName);
    });

    it('should throw an error when employee id already exists', async () => {
      const firstUser = await UserModel.create({
        first_name: faker.person.firstName(),
        last_name: faker.person.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await EmployeeUserModel.create({
        employee_id: 1,
        user: firstUser,
      }).save();

      await UserModel.create({
        first_name: faker.person.firstName(),
        last_name: faker.person.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await expect(
        userService.editUser(2, {
          first_name: faker.person.firstName(),
          employee_id: 1,
        }),
      ).rejects.toThrow('Employee ID already exists');
    });

    it('should create employee id', async () => {
      const user = await UserModel.create({
        first_name: faker.person.firstName(),
        last_name: faker.person.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await userService.editUser(user.id, {
        first_name: faker.person.firstName(),
        employee_id: 1,
      });

      const employeeUser = await EmployeeUserModel.findOne({
        where: { id: 1 },
        relations: ['user'],
      });

      expect(employeeUser).toBeDefined();
      expect(employeeUser.user.id).toBe(user.id);
    });

    it('should update employee id', async () => {
      const user = await UserModel.create({
        first_name: faker.person.firstName(),
        last_name: faker.person.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const oldEmployeeId = 1;
      const newEmployeeId = 2;

      const employeeUserRecord = await EmployeeUserModel.create({
        employee_id: oldEmployeeId,
        user: user,
      }).save();

      await userService.editUser(user.id, {
        first_name: faker.person.firstName(),
        employee_id: newEmployeeId,
      });

      const employeeUser = await EmployeeUserModel.findOne({
        where: { id: employeeUserRecord.id, employee_id: oldEmployeeId },
      });

      expect(employeeUser).toBeNull();

      const updatedEmployeeUser = await EmployeeUserModel.findOne({
        where: { id: employeeUserRecord.id, employee_id: newEmployeeId },
        relations: ['user'],
      });

      expect(updatedEmployeeUser).toBeDefined();
      expect(updatedEmployeeUser.user.id).toBe(user.id);
    });

    it('should throw an error when student id already exists', async () => {
      const firstUser = await UserModel.create({
        first_name: faker.person.firstName(),
        last_name: faker.person.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const studentId = 1;

      await StudentUserModel.create({
        student_id: studentId,
        user: firstUser,
      }).save();

      const newUser = await UserModel.create({
        first_name: faker.person.firstName(),
        last_name: faker.person.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await expect(
        userService.editUser(newUser.id, {
          first_name: faker.person.firstName(),
          student_id: studentId,
        }),
      ).rejects.toThrow('Student ID already exists');
    });

    it('should create student id', async () => {
      const user = await UserModel.create({
        first_name: faker.person.firstName(),
        last_name: faker.person.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await userService.editUser(user.id, {
        first_name: faker.person.firstName(),
        student_id: 1,
      });

      const studentUser = await StudentUserModel.findOne({
        where: { student_id: 1 },
        relations: ['user'],
      });

      expect(studentUser).toBeDefined();
      expect(studentUser.user.id).toBe(user.id);
    });

    it('should update student id', async () => {
      const user = await UserModel.create({
        first_name: faker.person.firstName(),
        last_name: faker.person.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const oldStudentId = 1;
      const newStudentId = 2;

      await StudentUserModel.create({
        student_id: oldStudentId,
        user: user,
      }).save();

      await userService.editUser(user.id, {
        first_name: faker.person.firstName(),
        student_id: newStudentId,
      });

      const updatedStudentUser = await StudentUserModel.findOne({
        where: { student_id: newStudentId },
        relations: ['user'],
      });

      expect(updatedStudentUser).toBeDefined();
      expect(updatedStudentUser.user.id).toBe(user.id);

      expect(await StudentUserModel.count()).toBe(1);
    });
  });

  describe('findUserCoursesById', () => {
    it('should return null when user not enrolled in any courses', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const result = await userService.findUserCoursesById(user.id);

      expect(result).toBeNull();
    });

    it('should return all courses found', async () => {
      const course = await CourseModel.create({
        course_code: 'COSC 499',
        course_name: 'Capstone Project',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        section_name: '001',
        invite_code: '123',
      }).save();

      const course2 = await CourseModel.create({
        course_code: 'MATH 101',
        course_name: 'Calculus I',
        created_at: 1_000_000_001,
        updated_at: 1_000_000_001,
        section_name: '002',
        invite_code: '456',
      }).save();

      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await CourseUserModel.create({
        course_role: CourseRoleEnum.STUDENT,
        user: { id: user.id },
        course: { id: course.id },
      }).save();

      await CourseUserModel.create({
        course_role: CourseRoleEnum.STUDENT,
        user: { id: user.id },
        course: { id: course2.id },
      }).save();

      const result = await userService.findUserCoursesById(user.id);
      expect(result).toHaveLength(2);
    });
  });
});
