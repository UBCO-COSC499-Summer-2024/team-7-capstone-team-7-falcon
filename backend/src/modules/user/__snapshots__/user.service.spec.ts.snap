// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService findOrCreateUser should create user 1`] = `
UserModel {
  "auth_type": "google_oauth",
  "avatar_url": "picture",
  "email": "john.doe@test.com",
  "first_name": "John",
  "id": 1,
  "last_name": "Doe",
  "password": null,
  "role": "student",
}
`;

exports[`UserService getUserById should return user 1`] = `
UserModel {
  "auth_type": "email",
  "avatar_url": null,
  "created_at": "1000000000",
  "email": "john.doe@test.com",
  "employee_user": EmployeeUserModel {
    "employee_id": 1,
    "id": 1,
  },
  "first_name": "John",
  "id": 1,
  "last_name": "Doe",
  "role": "student",
  "student_user": StudentUserModel {
    "student_id": 1,
  },
  "updated_at": "1000000000",
}
`;
